README.md
YouTube Downloader for Android
================================================================
originally by dentex (Samuele Rini)  
currently maintained by João Matos


Download YouTube video and extract or convert audio to mp3.
For Android 4+ 

*Free and without Ads*

################################################################

`versionCode="76"`

`versionName="3.5.5"`

DISCLAIMER
================================================================
    You, the user, accept the full responsibility for the usage of 
    this application, excluding the developer from any liability about 
    any consequence that may derive. Thus, this application has to be 
    considered as a mere technical service provider.
    
LICENSE
================================================================
Copyright (C) 2012-2013  Samuele Rini

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


OTHER LICENSES
================================================================
    Additional licenses informations about code used in this project
    is available from within the App's "About" menu:

    	Main Licenses 
     
    This App is released under the GPL-3.0. 
    (YouTubeDownloader for Android Copyright (C) 2012-2013 Samuele Rini) 
     
    The FFmpeg binary (http://www.ffmpeg.org/index.html), 
    used for the audio extraction feature, is released under the GPL-3.0. 
    (FFmpeg version 2.0.1 Copyright (C) 2000-2013 the FFmpeg developers) 
     
    The filechooser library is released under the MIT license. 
    (android-filechooser Copyright (C) 2012 Hai Bison) 
     
    The SlidingMenu library is released under the Apache License. 
    (Copyright (C) 2012 Jeremy Feinstein) 

    The liblame library, used to compile FFmpeg, is released under the LGPL. 
	A GitHub project, https://github.com/intervigilium/liblame, 
	has been used to compile the lame libraries for Android. 
    (liblame Copyright (C) 2010 Ethan Chen) 
    (LAME Copyright (C) 1999-2007 The LAME Project, 
    Copyright (C) 1999-2001 Mark Taylor, 
    Copyright (C) 1998 Micheal Cheng) 
    (mpglib Copyright (C) 1995-1997 Michael Hipp) 
     
    YTD`s mp3 tags are written with MyID3: a Java ID3 Tag Library 
    (http://www.fightingquaker.com/myid3 and 
    https://sites.google.com/site/eternalsandbox/myid3-for-android)
	released under the Apache license, as its dependency Jakarta Regexp library. 
    (MyID3 Copyright (C) 2008 Charles M. Chen, adapted for Android by Romulus Urakagi Ts`ai) 
     
    Picasso: Image downloading and caching library
	Copyright 2013 Square, Inc. - http://square.github.io/picasso/ 
     
    Launcher and notification icons by Sam Plane: http://samplane123.deviantart.com/, 
    released under the CC BY-NC-ND (http://creativecommons.org/licenses/by-nc-nd/3.0) 
     
    Credits for the custom download manager to Hiroshi Matsunaga (matsuhiro): 
    https://github.com/matsuhiro/AndroidDownloadManger (released "unlicensed"). 
     
     	Code excerpts Licenses 
     
    Stack Overflow (http://stackoverflow.com), a question and answer site for programmers where
	all the content is CC BY-SA 3.0 licensed (see http://creativecommons.org/licenses/by-sa/3.0) 
     
    checkMD5(...) and calculateMD5(...): 
    https://github.com/CyanogenMod/android_packages_apps_CMUpdater
	Copyright (C) 2012 The CyanogenMod Project, licensed under the GNU GPLv2 license 
     
    FfmpegController.java: GPL-3.0 - Copyright (C) 2009, Nathan Freitas, Orbot / The Guardian Project
	http://openideals.com/guardian - https://github.com/guardianproject/android-ffmpeg-java 
     
    Observer.java: reference: https://gist.github.com/shirou/659180 
     
    SectionedAdapter.java: GPL-3.0
	Copyright (C) 2008-2010 CommonsWare, LLC - portions Copyright (C) 2008 Jeffrey Sharkey 
     
    Utils.getCpuInfo(): http://www.roman10.net/how-to-get-cpu-information-on-android/ - by Liu Feipeng 
     
    Utils.scanMedia(...):
	http://www.grokkingandroid.com/adding-files-to-androids-media-library-using-the-mediascanner/
	by Wolfram Rittmeyer 
     
    DashboardAdapter.java and DashboardListItem.java: 
    code adapted from https://github.com/survivingwithandroid/Surviving-with-android/
	by Francesco Azzola (JFrankie) 
     
    Rhino.jar: open-source implementation of JavaScript written entirely in Java;
	https://developer.mozilla.org/en-US/docs/Rhino - license: http://www.mozilla.org/MPL/2.0/ 
     
    Javascript function `decryptSignature` from the Greasemonkey script 
    http://userscripts.org/scripts/show/25105 (MIT License) - by Gantt 
    
    QueueThread.java:
	reference: http://mindtherobot.com/blog/159/android-guts-intro-to-loopers-and-handlers/ 
    by Ivan Memruk 
     
     For more details, look for comments in YTD`s Java code. 
     
     	Translations 
    
    For more info go to: 
    www.getlocalization.com/ytdownloader 
    forum.xda-developers.com/showthread.php?p=37708791

    Note:
    the device-framed screenshots in the project's directory have 
    been generated with the "Device Frame Generator" Android App by 
    Prateek Srivastava, available at 
    <https://github.com/f2prateek/Device-Frame-Generator/>.
    The generated artwork is released ander the "Creative Commons 
    Attribution 3.0 Unported" license (CC BY).


CHANGELOG
================================================================
    
     v3.5.5 - Dec 21 2014 
    ----------------------------------- 
    [x] Add support for youtu.be links. 
    [x] Remove proprietary Bugsense library. 
     
     v3.5.4 - Dic 17 2013 
    ----------------------------------- 
    [x] possibility to import into Dashboard multiple files at once 
    [x] Dashboard bottom status bar 
     
     v3.5.3 - Dic 12 2013 
    ----------------------------------- 
    [x] auto-update bug fix 
     
     v3.5.2 - Dic 11 2013 
    ----------------------------------- 
    [x] fixed missing progress bars during mux 
    [x] Dashboard backup file can be renamed 
    [x] indicator for Dashboard background jobs 
     
     v3.5.1 - Dic 09 2013 
    ----------------------------------- 
    [x] fixed two small bugs 
     
     v3.5 - Dic 09 2013 
    ----------------------------------- 
    [x] Dashboard option to mux video-only and audio-only streams 
    [x] new FFmpeg builds from latest v2.1 (with NEON support) available for download 
    [x] option to choose between Settings or Dashboard as app entry-point-activity 
    [x] "Send" option into Dashboard items long-click menu 
    [x] all "in-progress" operations visible from Dashboard 
    [x] fixed Dark and Light themes appearance 
    [x] Android 4.4 target and many other optimizations 
    [x] Czech translation 
     
     v3.4.1 - Nov 19 2013 
    ----------------------------------- 
    [x] encrypted signature patch (7) 
    [x] layout fix for Samsung GT-P5110 and GT-N8000 devices 
     
     v3.4 - Nov 18 2013 
    ----------------------------------- 
    [x] video list filters via sliding menu 
    [x] encrypted signature patch (6) 
    [x] option to clear Dashboard into its menu (with checkbox to delete also data) 
    [x] removed support for MP4 1080p and FLV 480p "hidden" streams (not working anymore) 
     
     v3.3.1 - Nov 06 2013 
    ----------------------------------- 
    [x] fix for storage free space check 
     
     v3.3 - Nov 04 2013 
    ----------------------------------- 
    [x] MP4 1080p and FLV 480p "hidden" streams support 
    [x] again, better support for "secondary" YouTube streams 
    [x] fixed important bugs on resumed downloads and storage free space (thanks Christophelaures from XDA) 
     
     v3.2 - Oct 30 2013 
    ----------------------------------- 
    [x] better support for "secondary" YouTube streams 
    [x] support for audio files import into Dashboard 
    [x] other optimizations 
     
     v3.1 - Oct 26 2013 
    ----------------------------------- 
    [x] option to auto-enqueue FFmpeg tasks on downloads completion 
    [x] support for FLV video import into Dashboard 
    [x] support for video-only and audio-only "secondary" streams from YouTube (experimental) 
    [x] various bug fixes 
     
     v3.0.2 - Oct 11 2013 
    ----------------------------------- 
    [x] small fix 
     
     v3.0.1 - Oct 10 2013 
    ----------------------------------- 
    [x] layout fix for API 14,15,16 
    [x] other bug fixes 
     
     v3.0 - Oct 08 2013 
    ----------------------------------- 
    [x] Dashboard for interaction with downloaded video: 
     - show status/progress/speed 
     - click to: 
      ^ open video/audio file 
      ^ audio extraction/conversion 
     - long-click to manage files: 
      ^ copy 
      ^ move 
      ^ rename 
      ^ redownload 
      ^ remove 
      ^ delete 
      ^ pause/resume 
    [x] Custom download manager supporting: 
     - pause/resume 
     - YouTube link validation on resume through different networks and/or expire time 
    [x] new translations: 
     - Slovak 
     - Slovenian 
     - Vietnamese 
     - Finnish 
     
     v2.7 - Aug 20 2013 
    ----------------------------------- 
    [x] auto-patch experimental signature changes 
    [x] Android 4.3 target 
     
     v2.6.2 - Aug 08 2013 
    ----------------------------------- 
    [x] encrypted signature patch (5) 
     
     v2.6.1 - Jul 26 2013 
    ----------------------------------- 
    [x] option to show resolutions into the video list 
    [x] encrypted signature patch (4) 
     
     v2.6 - Jul 17 2013 
    ----------------------------------- 
    [x] fix for exp. signature parsing 
    [x] new translations: 
     - Arabic 
     - Danish 
     - Greek 
     
     v2.5.2 - Jul 14 2013 
    ----------------------------------- 
    [x] auto-fetch gantt\'s script to parse experimental signatures 
     
     v2.5.1 - Jul 12 2013 
    ----------------------------------- 
    [x] encrypted signature patch (3) 
     
     v2.5 - Jul 11 2013 
    ----------------------------------- 
    [x] encrypted signature patch (2) 
    [x] other small bugs fixed 
     
     v2.4 - Jul 02 2013 
    ----------------------------------- 
    [x] new launcher and notification icons by Sam Plane 
     
     v2.3.2 - Jun 28 2013 
    ----------------------------------- 
    [x] small bugs fixed 
     
     v2.3.1 - Jun 27 2013 
    ----------------------------------- 
    [x] encrypted signature patch 
     
     v2.3 - Jun 25 2013 
    ----------------------------------- 
    [x] encrypted signature support (thanks Gantt@userscripts.org) 
    [x] other bug/stability fixes 
     
     v2.2 - May 12 2013 
    ----------------------------------- 
    [x] fixed a bunch of BUUUUUGS !!! 
    [x] handles direct link clicks 
     
     v2.1 - May 11 2013 
    ----------------------------------- 
    [x] code optimizations 
    [x] mediaScanner bug fix (thanks Wolfram!) 
    [x] new translations: 
     - Chinese (China) 
     - Chinese (Hong Kong) 
     
     v2.0.6 - May 09 2013 
    ----------------------------------- 
    [x] dialog/toast when Downloads system app is not found 
    [x] Chinese (Taiwan) translation 
     
     v2.0.5 - May 09 2013 
    ----------------------------------- 
    [x] bug fix (FC when Downloads system app is not found)(3) 
     
     v2.0.4 - May 08 2013 
    ----------------------------------- 
    [x] bug fix (FC when Downloads system app is not found)(2) 
    [x] Hungarian translation 
     
     v2.0.3 - May 08 2013 
    ----------------------------------- 
    [x] bug fix (FC when Downloads system app is not found) 
     
     v2.0.2 - May 07 2013 
    ----------------------------------- 
    [x] bug fix 
     
     v2.0.1 - May 06 2013 
    ----------------------------------- 
    [x] bug fix 
     
     v2.0 - May 06 2013 
    ----------------------------------- 
    [x] audio extraction from downloaded video with optional conversion to mp3 (with FFmpeg) 
    [x] PayPal donation menu 
    [x] BugSense integration 
    [x] on the fly language switch 
    [x] fixed locale change to default 
    [x] 3 new option on SSH: 
     - use other ConnectBot forks 
     - use link inside same LAN 
     - SSH-send to long-press menu 
    [x] up-navigation for ativities 
    [x] nice translators list 
    [x] new translations: 
     - German 
     - Hebrew 
     - Polish (Poland) 
     - Portuguese (Brazil) 
     - Russian 
     - Spanish 
     - Western Farsi (Iran) 
     
     v1.9 - Mar 23 2013 
    ----------------------------------- 
    [x] YT web page parsing improved 
    [x] fixed notif. error when remove downloads not yet started 
    [x] 4K and 3D video support 
    [x] re-organized menus 
    [x] on the fly theme switch 
    [x] progress bar for video list 
    [x] option to enable logging 
    [x] Korean translation 
     
     v1.8.2 - Mar 15 2013 
    ----------------------------------- 
    [x] dark/light theme switcher 
    [x] option to show all the file sizes into the video list 
    [x] French translation 
     
     v1.8.1 - Mar 14 2013 
    ----------------------------------- 
    [x] notification bug fix 
     
     v1.8 - Mar 12 2013 
    ----------------------------------- 
    [x] extSdCard support 
    [x] long-press on video list item: menu to copy/share link 
    [x] auto update-check once a day 
    [x] new notification icon 
    [x] landscape orientation support 
    [x] YT video thumbnail preview 
    [x] entire row in v.list clickable 
    [x] Turkish translation 
    [x] Dutch translation 
     
     v1.7 - Mar 03 2012 
    ----------------------------------- 
    [x] translations: 
     - Italian 
     - Portuguese (Portugal) 
    [x] option to force the locale 
     
     v1.6 - Mar 01 2013 
    ----------------------------------- 
    [x] own YTD notification bar 
    [x] options for own and sys notif. 
    [x] added some credits 
     
     v1.5.2 - Feb 25 2013 
    ----------------------------------- 
    [x] better fix on bad video list 
    [x] remove stop download on click 
    [x] open DM on notif. bar click (still not working in CM ROM) 
     
     v1.5.1 - Feb 24 2013 
    ----------------------------------- 
    [x] fixed crash if no net is available 
    [x] fixed crash on bad video list 
     
     v1.5 - Feb 22 2013 
    ----------------------------------- 
    [x] online update check 
    [x] option to fetch filesizes 
    [x] stop download on click (maybe not working in CM 10.1) 
    [x] DownloadManager quick links 
    [x] settings and DM overflow menu 
    [x] reordered preferences 
    [x] small fixes 
     
     v1.4 - Jan 29 2013 
    ----------------------------------- 
    [x] licenses and code links 
    [x] Share this App option 
     
     v1.3.1 - Jan 19 2013 
    ----------------------------------- 
    [x] handle again http|https links (thanks Reiner!) 
     
     v1.3 - Jan 18 2013 
    ----------------------------------- 

NEWS.md
2014-12-21

Samuele Rini (dentex), the original author of this app, no longer
releases it as Free Software.
	
I, João Matos, will now be acting as the maintainer of this
maintenance fork of the last Free version of the original project.

I may consider developing new features for this app if I find the
interest to learn further about Android app developent, an activity I
have virtually no experience in at this point in time.

I will not be providing a binary package for this app for the time
being. Get it from [F-Droid](https://f-droid.org/repository/browse/?fdid=dentex.youtube.downloader).


dentex.youtube.downloader/src/dentex/youtube/downloader/YTD.java
/***
 	Copyright (c) 2012-2013 Samuele Rini
 	
	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	
	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
	GNU General Public License for more details.
	
	You should have received a copy of the GNU General Public License
	along with this program. If not, see http://www.gnu.org/licenses
	
	***
	
	https://github.com/dentex/ytdownloader/
	https://sourceforge.net/projects/ytdownloader/
	
	***
	
	Different Licenses and Credits where noted in code comments.
*/

package dentex.youtube.downloader;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import android.app.Activity;
import android.app.Application;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Environment;
import android.os.Handler;
import android.preference.Preference;
import android.support.v4.app.NotificationCompat;
import android.text.format.DateUtils;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.SparseArray;
import android.view.WindowManager;

import com.bugsense.trace.BugSenseHandler;

import dentex.youtube.downloader.queue.QueueThread;
import dentex.youtube.downloader.queue.QueueThreadListener;
import dentex.youtube.downloader.service.AutoUpgradeApkService;
import dentex.youtube.downloader.utils.PopUps;
import dentex.youtube.downloader.utils.Utils;

public class YTD extends Application implements QueueThreadListener{
	
	static String DEBUG_TAG = "YTD";
	public static Context ctx;

	// *** development configurations *** // TODO					// for release set to:
	// ================================== // ===========			// ==================
	public static String BugsenseApiKey = "00000000";				// actual api key
	public static boolean SHOW_ITAGS_AND_NO_SIZE_FOR_DUBUG = false;	// false
	
	private static String kudosKitKey = "0000000000";				// actual api key
	
	public static int SIG_HASH = -1892118308;   					// final string
	//public static int SIG_HASH = -118685648;  					// dev test: desktop
	//public static int SIG_HASH = 1922021506;  					// dev test: laptop
	// **********************************
	
	public static final String JSON_FILENAME = "dashboard.json";
	public static final String JSON_FILENAME_NO_EXT = "dashboard";
	public static final String JSON_FILENAME_EXT_ONLY = ".json";
	public static final String JSON_FOLDER = "json";
	public static File JSON_FILE = null;
	
	public static final String JSON_DATA_ID = "id";
	public static final String JSON_DATA_YTID = "ytid";
	public static final String JSON_DATA_POS = "pos";
	public static final String JSON_DATA_TYPE = "type";
	public static final String JSON_DATA_TYPE_V = "VIDEO";
	public static final String JSON_DATA_TYPE_V_O = "VIDEO-ONLY";
	public static final String JSON_DATA_TYPE_V_M = "VIDEO-MUX";
	public static final String JSON_DATA_TYPE_A_E = "AUDIO-EXTR";
	public static final String JSON_DATA_TYPE_A_M = "AUDIO-MP3";
	public static final String JSON_DATA_TYPE_A_O = "AUDIO-ONLY";
	public static final String JSON_DATA_STATUS = "status";
	public static final String JSON_DATA_STATUS_COMPLETED = "COMPLETED";
	public static final String JSON_DATA_STATUS_IN_PROGRESS = "IN_PROGRESS";
	public static final String JSON_DATA_STATUS_FAILED = "FAILED";
	public static final String JSON_DATA_STATUS_PAUSED = "PAUSED";
	public static final String JSON_DATA_STATUS_IMPORTED = "IMPORTED";
	public static final String JSON_DATA_STATUS_QUEUED = "QUEUED";
	public static final String JSON_DATA_PATH = "path";
	public static final String JSON_DATA_FILENAME = "filename";
	public static final String JSON_DATA_BASENAME = "basename";
	public static final String JSON_DATA_AUDIO_EXT = "audio_ext";
	public static final String JSON_DATA_SIZE = "size";
	
	public static final String ffmpegBinName = "ffmpeg";
	public static String FFMPEG_CURRENT_V = "_v2.1.1";
	
	public static String ARMv7a_NEON = "armv7a-neon";
	public static String ARMv7a = "armv7a";
//	public static String ARMv5te = "armv5te";
	public static String x86 = "x86";
	public static String UNSUPPORTED_CPU = "UNSUPPORTED";
	
	public static int _AUDIO_EXTR = 0;
	public static int _THUMB_EXTR = 1;
	
	public static SharedPreferences settings;
	public static SharedPreferences videoinfo;
	
	public static boolean isAnyAsyncInProgress = false;
	
	public static String PREFS_NAME = "dentex.youtube.downloader_preferences";
	public static String VIDEOINFO_NAME = "dentex.youtube.downloader_videoinfo";
	
	public static String pt1;
	public static String pt2;
	public static NotificationManager mNotificationManager;
	public static NotificationCompat.Builder mBuilder;
	public static List<Long> sequence = new ArrayList<Long>();
	
	//public static String USER_AGENT_FIREFOX = "Mozilla/5.0 (X11; Linux i686; rv:10.0) Gecko/20100101 Firefox/10.0";
	public static String USER_AGENT_FIREFOX = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:24.0) Gecko/20100101 Firefox/24.0";
	public static File dir_Downloads = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
	public static File dir_DCIM = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM);
	public static File dir_Movies = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MOVIES);
	
	public static Map<Long, Integer> mFFmpegPercentMap = new HashMap<Long, Integer>();

	public static final String THUMBS_FOLDER = "thumbs";
	public static double reduceFactor;
	
	public static QueueThread queueThread;
	public static Handler handler;
	
	public static final String VIEW_ALL_STRING = "";
	public static final int VIEW_ALL = -1;
	
	public static final int AO_FILTER = 10;
	public static final int VO_FILTER = 9;
	public static final int _3D_FILTER = 8;
	
	public static final int SD_FILTER = 7;
	public static final int MD_FILTER = 6;
	public static final int LD_FILTER = 5;
	public static final int HD_FILTER = 4;

	public static final int _3GP_FILTER = 3;
	public static final int FLV_FILTER = 2;
	public static final int WEBM_FILTER = 1;
	public static final int MP4_FILTER = 0;
	
	public static Integer[] iMp4 = { 18, 22, 37, 38, 59, 78, 82, 83, 84, 133, 134, 135, 136, 137, 138, 160, 264 };
	public static Integer[] iWebm = { 43, 44, 45, 46, 100, 101, 102, 242, 243, 244, 245, 246, 247, 248 };
	public static Integer[] iFlv = { 5, 6, 34, 35 };
	public static Integer[] i3gp = { 17, 36 };
	
	public static Integer[] iHd = { 22, 37, 38, 45, 46, 84, 102, 136, 137, 138, 247, 248, 264 };
	public static Integer[] iLd = { 35, 44, 59, 85, 135, 244, 245, 246 };
	public static Integer[] iMd = { 18, 34, 43, 78, 82, 100, 101, 134, 243 };
	public static Integer[] iSd = { 5, 6, 17, 36, 83, 133 };
	
	public static Integer[] i3d = { 82, 83, 84, 85, 100, 101, 102 };
	
	public static Integer[] iVo = { 133, 134, 135, 136, 137, 138, 160, 242, 243, 244, 245, 246, 247, 248, 264 };
	public static Integer[] iAo = { 139, 140, 141, 171, 172 };
	
	public static List<Integer> iVoList = Arrays.asList(iVo);
	public static List<Integer> iAoList = Arrays.asList(iAo);
	
	//kudos-kit
	public static final String APP_ID = "dentex.youtube.downloader";
		// The application key from the developer console
	public static final String APP_KEY = kudosKitKey;
		// The names of the in-app items associated with your application
		// You should use an empty array if you don't have in-app items to publish
	public static final String[] INAPP_IDS = new String[] {/*"ytd_kudos1", "ytd_kudos2", "ytd_kudos3"*/};
		// An array of feature strings; leave empty for now
	public static final String[] FEATURES = new String[] {};
		
	@Override
	public void onCreate() {
		Log.d(DEBUG_TAG, "onCreate");

		settings = getSharedPreferences(PREFS_NAME, 0);
		videoinfo = getSharedPreferences(VIDEOINFO_NAME, 0);
		
		BugSenseHandler.initAndStartSession(getApplicationContext(), BugsenseApiKey);
		
		queueThread = new QueueThread(this);
        queueThread.start();
        
        // Create the Handler. It will implicitly bind to the Looper
        // that is internally created for this thread (since it is the UI thread)
        handler = new Handler();
		
		//findProcessUid();
		
		ctx = getApplicationContext(); 
		JSON_FILE = new File(ctx.getDir(JSON_FOLDER, 0), JSON_FILENAME);
		
		detectSysDefLang();
		
		detectFirstLaunch();
		
		mBuilder =  new NotificationCompat.Builder(ctx);
		mNotificationManager = (NotificationManager) ctx.getSystemService(Context.NOTIFICATION_SERVICE);
		
		/*Log.i(DEBUG_TAG, 
				"\n --------------- " +
				Utils.getCpuInfo() + 
				"\n --------------- "
				+ "\nNeon support: " + Utils.neonCpu() + 
				"\n --------------- ");*/
		
		super.onCreate();
	}

	/*private void findProcessUid() {
		uid = android.os.Process.myUid();
		Log.d(DEBUG_TAG, "YTD's UID: " + uid);
	}*/
	
	private void detectFirstLaunch() {
		if (settings.getBoolean("first_launch", true)) {
			Log.i(DEBUG_TAG, "First launch for YTD!");
			settings.edit().putBoolean("first_launch", false).apply();
			reduceFactor = detectScreenDensity();
			
			JSON_FILE.delete();
			videoinfo.edit().clear().apply();
		} else {
			reduceFactor = Double.parseDouble(settings.getString("REDUCE_FACTOR", "1"));
			Log.d(DEBUG_TAG, "Retrieved a REDUCE_FACTOR of " + reduceFactor + " from prefs");
		}
	}
	
	private void detectSysDefLang() {
		String storedDefLang = settings.getString("DEF_LANG", "");
    	if (storedDefLang.isEmpty() && storedDefLang != null) {	
    		Locale defLocale = Locale.getDefault();
    		String defLang = defLocale.getLanguage();
    		Log.d(DEBUG_TAG, "Storing default system lang: " + defLang);
    		settings.edit().putString("DEF_LANG", defLang).commit();
    	}
	}

	private double detectScreenDensity() {
		DisplayMetrics displayMetrics = new DisplayMetrics();
	    ((WindowManager)getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay().getMetrics(displayMetrics);

	    double rf = 1;
		int density = displayMetrics.densityDpi;
		
		switch (density) {
	    case DisplayMetrics.DENSITY_HIGH: 
	    	rf = 1.44;
	    	break;
	    case DisplayMetrics.DENSITY_MEDIUM:
	    	rf = 2;
	    	break;
	    case DisplayMetrics.DENSITY_LOW:
	    	rf = 3;
	    }
	    Log.d(DEBUG_TAG, "DispalyDensity: " + density + " - storing a REDUCE_FACTOR of " + rf + " into prefs");
	    settings.edit().putString("REDUCE_FACTOR", String.valueOf(rf)).apply();;
	    return rf;
	}
	
	public static void updateInit(Activity act, boolean intoSettings, Preference up) {
		int prefSig = settings.getInt("APP_SIGNATURE", 0);
		Utils.logger("d", "prefSig: " + prefSig, DEBUG_TAG);
		
		if (prefSig == 0 ) {
			int currentHash = Utils.getSigHash(act);
			if (currentHash == SIG_HASH) {
				Utils.logger("d", "Found YTD signature: update check possile", DEBUG_TAG);
				if (intoSettings) up.setEnabled(true);
				
				if (settings.getBoolean("autoupdate", false)) {
					Utils.logger("i", "autoupdate enabled", DEBUG_TAG);
					autoUpdate();
				}
	    	} else {
	    		Utils.logger("d", "Found different signature: " + currentHash + " (F-Droid?). Update check cancelled.", DEBUG_TAG);
	    		if (intoSettings) {
	    			up.setEnabled(false);
	    			up.setSummary(R.string.update_disabled_summary);
	    		}
	    	}
			SharedPreferences.Editor editor = settings.edit();
	    	editor.putInt("APP_SIGNATURE", currentHash);
	    	if (editor.commit()) Utils.logger("d", "saving sig pref...", DEBUG_TAG);
		} else {
			if (prefSig == SIG_HASH) {
				Utils.logger("d", "YTD signature in PREFS: update check possile", DEBUG_TAG);
				if (intoSettings) up.setEnabled(true);
				
				if (settings.getBoolean("autoupdate", false)) {
					Utils.logger("i", "autoupdate enabled", DEBUG_TAG);
					autoUpdate();
				}
			} else {
				Utils.logger("d", "diffrent YTD signature in prefs (F-Droid?). Update check cancelled.", DEBUG_TAG);
				if (intoSettings) up.setEnabled(false);
			}
		}
	}
	
	public static void autoUpdate() {
        long storedTime = settings.getLong("time", 0);	// for release
        //long storedTime = 10000; 						// dev test: forces auto update
        
        boolean shouldCheckForUpdate = !DateUtils.isToday(storedTime);
        Utils.logger("i", "shouldCheckForUpdate: " + shouldCheckForUpdate, DEBUG_TAG);
        if (shouldCheckForUpdate) {
        	Intent intent = new Intent(ctx, AutoUpgradeApkService.class);
	        ctx.startService(intent);
        }
        
        long time = System.currentTimeMillis();
        if (settings.edit().putLong("time", time).commit()) Utils.logger("i", "time written in prefs", DEBUG_TAG);
	}

	public static void NoDownProvPopUp(Activity act) {
		PopUps.showPopUp(act.getString(R.string.error), act.getString(R.string.no_downloads_sys_app), "error", act);
	}
	
    public static void NotificationHelper(Context ctx) {
    	mBuilder =  new NotificationCompat.Builder(ctx); // to reset its DEFAULTS
    	
    	pt1 = ctx.getString(R.string.notification_downloading_pt1);
    	pt2 = ctx.getString(R.string.notification_downloading_pt2);
    	
    	Intent notificationIntent = new Intent(ctx, DashboardActivity.class);
    	notificationIntent.setFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
    	
    	PendingIntent contentIntent = PendingIntent.getActivity(ctx, 0, notificationIntent, 0);
    	
    	mBuilder.setSmallIcon(R.drawable.ic_stat_ytd)
			.setOngoing(true)
			.setContentTitle(ctx.getString(R.string.app_name))
			.setContentText(pt1 + " " + sequence.size() + " " + pt2)
			.setContentIntent(contentIntent);
    	
    	mNotificationManager.notify(1, mBuilder.build());
	}
    
    public static void removeIdUpdateNotification(long id) {
    	try {
	    	if (id != 0) {
				if (sequence.remove(id)) {
					Utils.logger("d", "ID " + id + " REMOVED from Notification", DEBUG_TAG);
				} else {
					Utils.logger("d", "ID " + id + " Already REMOVED from Notification", DEBUG_TAG);
				}
			} else {
				Utils.logger("w", "ID  not found!", DEBUG_TAG);
			}
	    	
	    	Utils.setNotificationDefaults(mBuilder);

			if (sequence.size() > 0) {
				mBuilder.setContentText(pt1 + " " + sequence.size() + " " + pt2)
						.setOngoing(true);
				mNotificationManager.notify(1, mBuilder.build());
			} else {
				mBuilder.setContentText(ctx.getString(R.string.notification_no_downloads))
						.setOngoing(false);
				mNotificationManager.notify(1, mBuilder.build());
				Utils.logger("d", "No downloads in progress.", DEBUG_TAG);
			}
		} catch (NullPointerException e) {
			Log.e(DEBUG_TAG, "NPE at removeIdUpdateNotification: " + e.getMessage());
			BugSenseHandler.sendExceptionMessage("NPE at removeIdUpdateNotification", e.getMessage(), e);
		}
	}

    @Override
	public void handleQueueThreadUpdate() {
    	
		handler.post(new Runnable() {
			@Override
			public void run() {
				int total = queueThread.getTotalQueued();
				int completed = queueThread.getTotalCompleted();

				Utils.logger("i", String.format("Auto audio extractions completed: "
						+ "%d of %d", completed, total), DEBUG_TAG);
				
				if (completed == total) {
					queueThread.resetQueue();
					queueThread.pushNotificationText(ctx, ctx.getString(R.string.auto_audio_extr_completed), false);
				} else {
					queueThread.pushNotificationText(ctx, ctx.getString(R.string.auto_audio_extr_progress, completed, total), true);
				}
			}
		});
	}

	public static CharSequence getListFilterConstraint(int c) {
		//0
		List<Integer> iMp4List = Arrays.asList(iMp4);
		//1
		List<Integer> iWebmList = Arrays.asList(iWebm);
		//2
		List<Integer> iFlvList = Arrays.asList(iFlv);
		//3
		List<Integer> i3gpList = Arrays.asList(i3gp);
		
		//4
		List<Integer> iHdList = Arrays.asList(iHd);
		//5
		List<Integer> iLdList = Arrays.asList(iLd);
		//6
		List<Integer> iMdList = Arrays.asList(iMd);
		//7
		List<Integer> iSdList = Arrays.asList(iSd);
		
		//8
		List<Integer> i3dList = Arrays.asList(i3d);
		
		//9
		List<Integer> iVoList = Arrays.asList(iVo);
		//10
		List<Integer> iAoList = Arrays.asList(iAo);
		
		SparseArray<List<Integer>> filtersMap = new SparseArray<List<Integer>>();
		
		filtersMap.put(MP4_FILTER, iMp4List);
		filtersMap.put(WEBM_FILTER, iWebmList);
		filtersMap.put(FLV_FILTER, iFlvList);
		filtersMap.put(_3GP_FILTER, i3gpList);
		filtersMap.put(HD_FILTER, iHdList);
		filtersMap.put(LD_FILTER, iLdList);
		filtersMap.put(MD_FILTER, iMdList);
		filtersMap.put(SD_FILTER, iSdList);
		filtersMap.put(_3D_FILTER, i3dList);
		filtersMap.put(VO_FILTER, iVoList);
		filtersMap.put(AO_FILTER, iAoList);
		
		if (c == -1) return VIEW_ALL_STRING;
		
		CharSequence constraint = null;
		List<Integer> selectedMap = filtersMap.get(c);
		
		for (int i = 0; i < selectedMap.size(); i++) {
			if (constraint == null) { 
				constraint = String.valueOf(selectedMap.get(i));
			} else {
				constraint = constraint + "/" + selectedMap.get(i);
			}
		}
		//Utils.logger("i", "ListFilterConstraint: " + constraint, DEBUG_TAG);
		return constraint;
	}
	
	public static CharSequence getMultipleListFilterConstraints(int[] c) {
		CharSequence constraint = null;
		for (int i = 0 ; i < c.length; i++) {
			if (constraint == null) { 
				constraint = getListFilterConstraint(c[i]);
			} else {
				constraint = constraint + "/" + getListFilterConstraint(c[i]);
			}
		}
		return constraint;
	}
}
